src/Psalm/Type/Atomic/TTemplateKeyOf.php

```diff
@@ -2,12 +2,16 @@
 
 namespace Psalm\Type\Atomic;
 
+use Psalm\Codebase;
+use Psalm\Internal\Type\TemplateInferredTypeReplacer;
+use Psalm\Internal\Type\TemplateResult;
+use Psalm\Type\Atomic;
 use Psalm\Type\Union;
 
 /**
- * Represents the type used when using TKeyOfClassConstant when the type of the class constant array is a template
+ * Represents the type used when using TKeyOfArray when the type of the array is a template
  */
-class TTemplateKeyOf extends TArrayKey
+class TTemplateKeyOf extends Atomic
```

Comment:
why this change? It seems like it would indeed be an array-key?

Reply:
This caused the bug with nested templates. `TArrayKey` extends from `Scalar`, so if `TTemplateKeyOf` have been compared to a Scalar value the `ScalarTypeComparator` has been called. But if it has been compared to a `TTemplateParam` the `AtomicTypeComparator` has been called. This lead to a duplicate implementation of the `isContainedBy`-logic. Therefor I decided to extend from `Atomic` as there is no benefit from extending from `Scalar`.

Reply:
ok, seems fine